# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/ContentParent.cpp
# Commit: 360b12c6b3e4
# Full Hash: 360b12c6b3e4e74203ba90324580a6c5e149b66b
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2019-07-16 04:07:00
# Regressor Bug: 1563604
# File Overlap Count: 1
# Description:
#   Bug 1563604 - Part 1: Confirm detach calls before dropping refs, r=peterv
#   
#   This ensures that the object is not discarded while there are still in-flight
#   messages from the parent process, and vice-versa, which could cause crashes if a
#   CC was run before any pending messages have arrived.
# ==============================================================================

diff -r 80fff55bee77 -r 360b12c6b3e4 dom/ipc/ContentParent.cpp
--- a/dom/ipc/ContentParent.cpp	Mon Jul 15 17:19:17 2019 +0000
+++ b/dom/ipc/ContentParent.cpp	Mon Jul 15 20:44:50 2019 +0000
@@ -5817,14 +5817,21 @@
 }
 
 mozilla::ipc::IPCResult ContentParent::RecvDetachBrowsingContext(
-    BrowsingContext* aContext) {
-  if (!aContext) {
+    uint64_t aContextId, DetachBrowsingContextResolver&& aResolve) {
+  // NOTE: Immediately resolve the promise, as we've received the message. This
+  // will allow the content process to discard references to this BC.
+  aResolve(true);
+
+  // NOTE: It's OK if we don't have this context anymore. It was just already
+  // detached, return.
+  RefPtr<BrowsingContext> context = BrowsingContext::Get(aContextId);
+  if (!context) {
     MOZ_LOG(BrowsingContext::GetLog(), LogLevel::Debug,
             ("ParentIPC: Trying to detach already detached"));
     return IPC_OK();
   }
 
-  if (!aContext->Canonical()->IsOwnedByProcess(ChildID())) {
+  if (!context->Canonical()->IsOwnedByProcess(ChildID())) {
     // We're trying to detach a child BrowsingContext in another child
     // process. This is illegal since the owner of the BrowsingContext
     // is the proccess with the in-process docshell, which is tracked
@@ -5833,14 +5840,18 @@
 
     MOZ_LOG(BrowsingContext::GetLog(), LogLevel::Warning,
             ("ParentIPC: Trying to detach out of process context 0x%08" PRIx64,
-             aContext->Id()));
+             context->Id()));
     return IPC_OK();
   }
 
-  aContext->Detach(/* aFromIPC */ true);
-
-  aContext->Group()->EachOtherParent(this, [&](ContentParent* aParent) {
-    Unused << aParent->SendDetachBrowsingContext(aContext);
+  context->Detach(/* aFromIPC */ true);
+
+  context->Group()->EachOtherParent(this, [&](ContentParent* aParent) {
+    // Hold a reference to `context` until the response comes back to ensure it
+    // doesn't die while messages relating to this context are in-flight.
+    auto resolve = [context](bool) {};
+    auto reject = [context](ResponseRejectReason) {};
+    aParent->SendDetachBrowsingContext(context->Id(), resolve, reject);
   });
 
   return IPC_OK();